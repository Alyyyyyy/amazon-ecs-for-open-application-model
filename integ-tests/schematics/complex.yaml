apiVersion: core.oam.dev/v1alpha1
kind: ComponentSchematic
metadata:
  name: nginx-replicated
  labels:
    app: my-nginx-replicated-app
  annotations:
    version: "1.0.1"
    description: A worker that runs nginx
spec:
  workloadType: core.oam.dev/v1alpha1.Worker
  osType: linux
  arch: arm64
  containers:
    - name: server
      image: nginx:latest
      cmd:
        - nginx
      args:
        - "--debug"
      config:
        - path: "/etc/access/default_user.txt"
          value: "admin"
        - path: "/var/run/db-data"
          fromParam: "sourceData"
      resources:
        cpu:
          required: 4
        memory:
          required: 10G
        gpu:
          required: 1.0
        volumes:
          - name: "configuration"
            mountPath: /etc/config
            accessMode: RO
            sharingPolicy: Shared
            disk:
              required: "2G"
              ephemeral: n
      livenessProbe:
        exec:
          command:
            - "wget"
            - "-qO"
            - "-"
            - "http://localhost"
        initialDelaySeconds: 5
        periodSeconds: 12
        timeoutSeconds: 3
        successThreshold: 2
        failureThreshold: 4
---
apiVersion: core.oam.dev/v1alpha1
kind: ComponentSchematic
metadata:
  name: nginx-replicated-server
  labels:
    app: my-nginx-replicated-app-server
  annotations:
    version: "1.0.1"
    description: A worker that runs nginx
spec:
  workloadType: core.oam.dev/v1alpha1.Server
  osType: linux
  arch: arm64
  containers:
    - name: server
      image: nginx:latest
      cmd:
        - nginx
      args:
        - "--debug"
      config:
        - path: "/etc/access/default_user.txt"
          value: "admin"
        - path: "/var/run/db-data"
          fromParam: "sourceData"
      resources:
        cpu:
          required: 2
        memory:
          required: 14G
        gpu:
          required: 2.0
        volumes:
          - name: "configuration"
            mountPath: /etc/config
            accessMode: RO
            sharingPolicy: Shared
            disk:
              required: "2G"
              ephemeral: n
      ports:
        - name: http
          containerPort: 8080
          protocol: tcp
      livenessProbe:
        httpGet:
          port: 8081
          path: /ok
        initialDelaySeconds: 5
        periodSeconds: 12
        timeoutSeconds: 3
        successThreshold: 2
        failureThreshold: 4
      readinessProbe:
        httpGet:
          port: 8084
          path: /ok-ready
        initialDelaySeconds: 1
        periodSeconds: 2
        timeoutSeconds: 4
        successThreshold: 3
        failureThreshold: 5
---
apiVersion: core.oam.dev/v1alpha1
kind: ApplicationConfiguration
metadata:
  name: complex-example
  annotations:
    version: v1.0.0
    description: "Complex worker and server example"
spec:
  variables:
  components:
    - componentName: nginx-replicated-server
      instanceName: web-front-end
    - componentName: nginx-replicated
      instanceName: backend
