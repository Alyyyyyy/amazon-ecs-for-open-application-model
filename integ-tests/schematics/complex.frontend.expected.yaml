AWSTemplateFormatVersion: 2010-09-09
Description: Amazon ECS infrastructure for complex-example web-front-end

Resources:
  LogGroup:
    Type: AWS::Logs::LogGroup
    Properties:
      LogGroupName: oam-ecs-complex-example-web-front-end

  TaskDefinition:
    Type: AWS::ECS::TaskDefinition
    Properties:
      Family: oam-ecs-complex-example-web-front-end
      NetworkMode: awsvpc
      RequiresCompatibilities:
        - FARGATE
      Cpu: 2.00 vcpu
      Memory: '14336'
      ExecutionRoleArn: !GetAtt ExecutionRole.Arn
      ContainerDefinitions:
        - Name: server
          Image: nginx:latest
          ResourceRequirements:
            - Type: GPU
              Value: '2'
          MountPoints:
            - ContainerPath: /etc/config
              ReadOnly:  true
              SourceVolume: configuration
          EntryPoint:
            - "nginx"
          Command:
            - "--debug"
          PortMappings:
            - ContainerPort: 8080
              Protocol:  tcp
          LogConfiguration:
            LogDriver: awslogs
            Options:
              awslogs-region: !Ref AWS::Region
              awslogs-group: !Ref LogGroup
              awslogs-stream-prefix: oam-ecs
      Volumes:
        - Name: configuration

  ExecutionRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Effect: Allow
            Principal:
              Service: ecs-tasks.amazonaws.com
            Action: 'sts:AssumeRole'

      ManagedPolicyArns:
        - !Sub 'arn:${AWS::Partition}:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy'

  ContainerSecurityGroup:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupDescription: oam-ecs-complex-example-web-front-end-ContainerSecurityGroup
      VpcId:
        Fn::ImportValue: oam-ecs-VpcId

  Service:
    Type: AWS::ECS::Service
    Properties:
      Cluster:
        Fn::ImportValue: oam-ecs-ECSCluster
      TaskDefinition: !Ref TaskDefinition
      DeploymentConfiguration:
        MinimumHealthyPercent: 100
        MaximumPercent: 200
      DesiredCount: 1
      LaunchType: FARGATE
      NetworkConfiguration:
        AwsvpcConfiguration:
          AssignPublicIp: DISABLED
          Subnets:
            Fn::Split:
              - ','
              - Fn::ImportValue: oam-ecs-PrivateSubnets
          SecurityGroups:
            - !Ref ContainerSecurityGroup
      LoadBalancers:
        - ContainerName: server
          ContainerPort: 8080
          TargetGroupArn: !Ref TargetGroupServer8080
      HealthCheckGracePeriodSeconds: 5
    DependsOn:
      - LBListenerServer8080



  SGLoadBalancerToContainers:
    Type: AWS::EC2::SecurityGroupIngress
    Properties:
      Description: Ingress from anywhere on the internet through the public NLB
      GroupId: !Ref ContainerSecurityGroup
      IpProtocol: '-1'
      CidrIp: 0.0.0.0/0

  PublicLoadBalancer:
    Type: AWS::ElasticLoadBalancingV2::LoadBalancer
    Properties:
      Type: network
      Scheme: internet-facing
      Subnets:
        Fn::Split:
          - ','
          - Fn::ImportValue: oam-ecs-PublicSubnets

  LBListenerServer8080:
    Type: AWS::ElasticLoadBalancingV2::Listener
    Properties:
      DefaultActions:
        - TargetGroupArn: !Ref TargetGroupServer8080
          Type: 'forward'
      LoadBalancerArn: !Ref 'PublicLoadBalancer'
      Port: 8080
      Protocol:  TCP

  TargetGroupServer8080:
    Type: AWS::ElasticLoadBalancingV2::TargetGroup
    Properties:
      Protocol:  TCP
      TargetType: ip
      Port: 8080
      VpcId:
        Fn::ImportValue: oam-ecs-VpcId
      TargetGroupAttributes:
      - Key: deregistration_delay.timeout_seconds
        Value: '30'

      HealthCheckProtocol: HTTP
      HealthCheckPath: /ok
      HealthCheckPort: '8081'
      HealthCheckTimeoutSeconds:  3

      HealthCheckIntervalSeconds:  12
      HealthyThresholdCount:  2
      UnhealthyThresholdCount:  4



Outputs:
  CloudFormationStackConsole:
    Description: The AWS console deep-link for the CloudFormation stack
    Value: !Sub https://console.aws.amazon.com/cloudformation/home?region=${AWS::Region}#/stacks/stackinfo?stackId=${AWS::StackName}

  ECSServiceConsole:
    Description: The AWS console deep-link for the ECS service
    Value: !Sub https://console.aws.amazon.com/ecs/home?region=${AWS::Region}#/clusters/oam-ecs/services/${Service.Name}

  ServerPort8080Endpoint:
    Description: The endpoint for container Server on port 8080
    Value: !Sub '${PublicLoadBalancer.DNSName}:8080'

